sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java

```diff
@@ -1170,4 +887,344 @@ private void waitForFuturesToFinish() throws ExecutionException, InterruptedExce
+    }
+
+    /** Specify the password used for authentication. */
+    public ReadAll<T> withPassword(String password) {
+      checkArgument(password != null, "password can not be null");
+      return withPassword(ValueProvider.StaticValueProvider.of(password));
+    }
+
+    /** Specify the password used for authentication. */
+    public ReadAll<T> withPassword(ValueProvider<String> password) {
+      return builder().setPassword(password).build();
+    }
+
+    /** Specify the local DC used for the load balancing. */
+    public ReadAll<T> withLocalDc(String localDc) {
+      checkArgument(localDc != null, "localDc can not be null");
+      return withLocalDc(ValueProvider.StaticValueProvider.of(localDc));
+    }
+
+    /** Specify the local DC used for the load balancing. */
+    public ReadAll<T> withLocalDc(ValueProvider<String> localDc) {
+      return builder().setLocalDc(localDc).build();
+    }
+
+    public ReadAll<T> withConsistencyLevel(String consistencyLevel) {
+      checkArgument(consistencyLevel != null, "consistencyLevel can not be null");
+      return withConsistencyLevel(ValueProvider.StaticValueProvider.of(consistencyLevel));
+    }
+
+    public ReadAll<T> withConsistencyLevel(ValueProvider<String> consistencyLevel) {
+      return builder().setConsistencyLevel(consistencyLevel).build();
+    }
+
+    public ReadAll<T> withGroupingFn(SerializableFunction<RingRange, Integer> groupingFunction) {
+      return builder().setGroupingFn(groupingFunction).build();
+    }
+
+    public ReadAll<T> withSplitCount(ValueProvider<Integer> splitCount) {
+      return builder().setSplitCount(splitCount).build();
+    }
+
+    public ReadAll<T> withSplitCount(Integer splitCount) {
+      checkArgument(splitCount != null, "splitCount can not be null");
+      return withSplitCount(ValueProvider.StaticValueProvider.<Integer>of(splitCount));
+    }
+
+    /**
+     * A factory to create a specific {@link Mapper} for a given Cassandra Session. This is useful
+     * to provide mappers that don't rely in Cassandra annotated objects.
+     */
+    public ReadAll<T> withMapperFactoryFn(SerializableFunction<Session, Mapper> mapperFactory) {
+      checkArgument(
+          mapperFactory != null,
+          "CassandraIO.withMapperFactory" + "(withMapperFactory) called with null value");
+      return builder().setMapperFactoryFn(mapperFactory).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PCollection<Read<T>> input) {
+      checkArgument((hosts() != null && port() != null), "WithHosts() and withPort() are required");
+      checkArgument(keyspace() != null, "withKeyspace() is required");
+      checkArgument(table() != null, "withTable() is required");
+      checkArgument(entity() != null, "withEntity() is required");
+      checkArgument(coder() != null, "withCoder() is required");
+      checkArgument(groupingFn() != null, "GroupingFn OR splitCount must be set");
+      try (Cluster cluster =
```

Comment:
The key point here is to apply for every read the split function and produce subsequent reads with RingRanges that then are Reshuffled and passed to a ParDo with the ReadFn function.

Reply:
Not sure I understand what you mean here.  I'm doing a shuffle by key here, `FlattenGrouped` here is what I called it, following the other connectors general pattern.